[SR-12460] Don't crash when missing a type in a Diagnostic message.

blueeor opened the pull request:
Print an empty string for the missing type.

Resolves SR-12460

Given this swift file:

```
extension {
  func ==(lhs: Foo, rhs: Foo) -> Bool {}
} 
```

And the swift arguments that XCode uses to compile swift.

Produce this output:

```
/Users/eyork/dev/test/Test/Test/main.swift:13:11: error: expected type name in extension declaration
extension {
          ^
/Users/eyork/dev/test/Test/Test/main.swift:14:8: **error: operator '==' declared in extension of '' must be 'static'**
  func ==(lhs: Foo, rhs: Foo) -> Bool {}
       ^
  static 
/Users/eyork/dev/test/Test/Test/main.swift:14:16: error: use of undeclared type 'Foo'
  func ==(lhs: Foo, rhs: Foo) -> Bool {}
               ^~~
/Users/eyork/dev/test/Test/Test/main.swift:14:26: error: use of undeclared type 'Foo'
  func ==(lhs: Foo, rhs: Foo) -> Bool {}
                         ^~~
Program ended with exit code: 1

```
This line in bold has the fixed result. It will refer to the missing type as ''.
Before this fix, the compiler would crash when producing that line.

blueeor committed:
Don't crash when missing a type in a Diagnostic message.
Print an empty string for the missing type.

Resolves SR-12460

xedin reviewed (changes requested):
Thank you for the contribution! I'd suggest we actually fix declaration checker to verify that extension has a name before trying to produce this diagnostic about operator and add an `assert` in DiagnosticEngine to make sure that `TypeRepr *` is not null.

Diagnostic for `non-static operator declaration` is emitted here https://github.com/apple/swift/blob/master/lib/Sema/TypeCheckDecl.cpp#L679-L682. I think it might be worth adding a new diagnostic message which doesn't require a name and emit that instead if situations like this.

blueeor commented:
Ok, I will make those updates.

blueeor committed:
Added new diagnostic message, eg. "operator '==' declared in extension must be 'static'"
Added assert for missing a type name
IsStaticRequest::evaluate now checks for ED->getExtendedTypeRepr().

blueeor commented:
Made that update. The new diagnostic message is "operator '==' declared in extension must be 'static''. How does that look?

blueeor committed:
Format update

blueeor committed:
Moved if check

AnthonyLatsis commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -904,6 +904,9 @@ ERROR(nonstatic_operator_in_nominal,none,
       (Identifier, TypeRepr*))
+ERROR(nonstatic_operator_in_extension_no_type,none,
+"operator %0 declared in extension must be 'static'",
+(Identifier))
```
Maybe incorporating this into the old diagnostic using `%select` is a better idea? We can them pass a bool rather than duplicating the diagnosing code.

blueeor replied:
Updated to use one diagnostic.

AnthonyLatsis commented on lib/AST/DiagnosticEngine.cpp:
```diff
@@ -585,10 +585,12 @@ static void formatDiagnosticArgument(StringRef Modifier,
     break;
-  case DiagnosticArgumentKind::PatternKind:
+
+ case DiagnosticArgumentKind::PatternKind:
```
Accidental indentation change?

blueeor replied:
Yes, I will fix that.

AnthonyLatsis commented on lib/AST/DiagnosticEngine.cpp:
```diff
@@ -585,10 +585,12 @@ static void formatDiagnosticArgument(StringRef Modifier,
   }
   case DiagnosticArgumentKind::TypeRepr:
     assert(Modifier.empty() && "Improper modifier for TypeRepr argument");
+    assert(Arg.getAsTypeRepr() && "TypeRepr argument has no name");
```
Since the argument is a pointer, I think `TypeRepr argument is null` would be more accurate.

blueeor replied:
Updated.

AnthonyLatsis reviewed (commented):
Good catch! Could you please add a test so that we don't regress again? `swift/test/decl/ext/extensions.swift` is a good spot.

AnthonyLatsis commented on lib/Sema/TypeCheckDecl.cpp:
```diff
@@ -676,10 +676,16 @@ IsStaticRequest::evaluate(Evaluator &evaluator, FuncDecl *decl) const {
+            decl->diagnose(diag::nonstatic_operator_in_extension_no_type, operatorName)
+                .fixItInsert(decl->getAttributeInsertionLoc(/*forModifier=*/true),
+                             "static ");
+        }
```
The convention is 2 spaces for the indent width (not 4).

blueeor replied:
Fixed

theblixguy replied:
Could you also run clang-format on your changes to make sure everything is properly formatted and indented?

blueeor replied:
Will do

blueeor committed:
Updated alert message to "TypeRepr argument is null"
Formatting updates

blueeor committed:
Moved diagnostic to one message

blueeor committed:
Add test for extention without type

blueeor commented:
I added a test in `swift/test/decl/ext/extensions.swift` but I am not sure how to run that single test.

AnthonyLatsis commented on lib/Sema/TypeCheckDecl.cpp:
```diff
@@ -676,8 +676,9 @@ IsStaticRequest::evaluate(Evaluator &evaluator, FuncDecl *decl) const {
-      decl->diagnose(diag::nonstatic_operator_in_extension,
-                     operatorName, ED->getExtendedTypeRepr())
+      decl->diagnose(diag::nonstatic_operator_in_extension, operatorName,
+                     ED->getExtendedTypeRepr() != NULL,
```
Please use `nullptr` or `(bool) ED->getExtendedTypeRepr()`

blueeor replied:
Fixed. BTW: What is the convention here for resolving conversations?

AnthonyLatsis replied:
There is no convention really, this is just mostly used as an organisational tool. Usually, I resolve a conversation if someone addressed my suggestion or I myself have addressed a suggestion as an author, given there is some confidence that the conversation has exhausted itself and the underlying issue/idea/etc is trivial enough to fold away from the direct line of sight of other reviewers.

AnthonyLatsis reviewed (commented):
Looks good, thank you! Have a look at https://github.com/apple/swift/blob/master/docs/Testing.md#using-litpy for how to use `lit` to run the tests. `--filter=` — to run a single test file.

blueeor committed:
Use nullptr

AnthonyLatsis reviewed (commented):
I have one final suggestion, the rest is up to @xedin.

blueeor committed:
Update test comment.

blueeor commented on test/decl/ext/extensions.swift:
```diff
@@ -38,6 +38,11 @@ func nestingTest6() {
   extension Foo {} // expected-error {{declaration is only valid at file scope}}
 }
 
+//===--- Test extensions without type and a member that should be static.
```
ok, I updated the description.

xedin reviewed (approved):
LGTM, thank you!

xedin commented:
@swift-ci please smoke test

xedin commented:
@blueeor Looks like there are test failures.

blueeor commented:
ok, I will look at that.

blueeor commented:
The test fails for swift 4:
`RUN: at line 1';   /Users/eyork/apple-swift/build/Ninja-DebugAssert/swift-macosx-x86_64/bin/swiftc -frontend -target x86_64-apple-macosx10.9  -module-cache-path '/Users/eyork/apple-swift/build/Ninja-DebugAssert/swift-macosx-x86_64/swift-test-results/x86_64-apple-macosx10.9/clang-module-cache' -sdk '/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk'  -swift-version 4  -ignore-module-source-info -typo-correction-limit 10  -typecheck -verify -disable-objc-attr-requires-foundation-module /Users/eyork/apple-swift/swift/test/decl/ext/extensions.swift`

I am assuming that it's working correctly for swift 5.

I don't know how to set the error test in extensions.swift to only be for swift 5?

xedin commented:
Why does it fail for Swift 4 and not 5?

AnthonyLatsis commented:
@blueeor Have you tried running `BUILD_DIR/bin/swift `on a separate file with the relevant test code to verify the correct diagnostic is emitted?

blueeor commented:
I was mistaken. This doesn't have anything to do with swift language 4 or 5.

This test is running swiftc with different options vs when xcode runs swift. With the default options on swift, this diagnosis message doesn't appear.  I haven't yet figured out which swift option will make this diagnosis message appear (a pointer in the right direction would help me). The other question is would I need to put this test in it's own file and add a test that runs swiftc with the correct options? Or is there already a test that runs swift with the same options that XCode would use?

xedin commented:
@blueeor I think it would be fine to add a new test file with particular set of flags in RUN which reproduce this problem, you can use SR number as a file name, we do that in test-suite already.

AnthonyLatsis commented:
The code surrounding the emission of the error is not showing any dependence on an invocation flag, and the diagnostic *does* appear with a proper extension, so it has to be something else.

LucianoPAlmeida commented:
> I haven't yet figured out which swift option will make this diagnosis message appear (a pointer in the right direction would help me). 
>

@blueeor From the stack-trace seems like the crash appears when trying to index symbols (`IndexSwiftAST`), so you can pass something like `target-swift-ide-test -print-indexed-symbols` 
could be that @xedin?

AnthonyLatsis commented:
OK, so in between recent snapshots and the Xcode 11.3 release something happened that caused us to completely skip validation of members inside these broken extensions, which is why the error is not emitted normally in this case. We either have to fix that or hit the code path by using the appropriate command to emit index data (@LucianoPAlmeida pointed out the right one).

blueeor commented:
@LucianoPAlmeida Thanks. I will update the test with that or I guess make a new test for this. 

@AnthonyLatsis It sounds like you are saying there is a larger bug at play here?

blueeor committed:
Fixed test to add target-swift-ide-test -print-indexed-symbols

blueeor commented:
I updated the tests and it passes for me now. I pushed the commit. Will the ci system pick this up or does it need a prompt to run again?

LucianoPAlmeida commented:
> I updated the tests and it passes for me now. I pushed the commit. Will the ci system pick this up or does it need a prompt to run again?

@blueeor Someone with the rights should kick the CI
cc @xedin :)

AnthonyLatsis commented on test/decl/ext/extensions_no_type.swift:
```diff
@@ -0,0 +1,8 @@
+//===--- Test that we don't crash when validating members inside an extension with no type name.
+extension { // expected-error {{expected type name in extension declaration}}
+  static func ==(lhs: Foo, rhs: Foo) -> Bool {} // expected-error {{operator '==' declared in extension must be 'static'}}
+}
```
I think we should move the test to `test/Index` now that we use this command.
Also, since the test is somewhat unusual, can you name it after the SR to provide more context? This will make it easier to move back to `decl/ext` once we fix the validation issue.

AnthonyLatsis replied:
Oh no, the crash is guarded, but the `expected-error` checks here will have no effect. The only way to test that your error is actually there is to `CHECK:` for it in the command output. That's a pretty dirty workaround, so let's wait for what Pavel has to say.

blueeor replied:
I renamed the the test file to SR-12460.swift.

I don't understand your comment on <CHECK:>, the test seems to work correctly with those compile options? What I am missing?

AnthonyLatsis replied:
> I don't understand your comment on CHECK:, the test seems to work correctly with those compile options? What I am missing?

The test will still pass if you mess up the expected errors. You need `%target-typecheck-verify-swift` for the `expected-error` check to take effect, but the error you want is only emitted as a byproduct of the `%target-swift-ide-test -print-indexed-symbols` output.

xedin replied:
I don’t think typecheck-verify command is really necessary here, this output could be verified with FileCheck. I’ll make some experiments tomorrow and let you know.

blueeor committed:
Updated test file name

blueeor committed:
Updated test file name

xedin commented:
Here is the test-case we can use for this issue, let's have it as `test/decl/ext/sr_12460.swift`

```
// RUN: %target-swift-ide-test -print-indexed-symbols -source-filename %s 2>&1 | %FileCheck -check-prefix CHECK %s

//Test that we don't crash when validating members inside an extension with no type name.

// CHECK: :[[@LINE+1]]:11: error: expected type name in extension declaration
extension {
  // CHECK: :[[@LINE+1]]:8: error: operator '==' declared in extension must be 'static'
  func ==(lhs: Any, rhs: Any) -> Bool {}
}
```

blueeor committed:
Update test file

blueeor commented:
@xedin I updated the test file with those changes.

xedin commented:
@swift-ci please smoke test

xedin commented:
Thank you, @blueeor! Please don't forget to resolve associated SR.
